autogen.pl

```diff
@@ -1149,6 +1149,105 @@ sub in_tarball {
     return $tarball;
 }
 
+##############################################################################
+
+my %cached_config_timestamps;
+
+sub replace_config_sub_guess {
+    # This could be simpler if we could use some Perl modules for this
+    # functionality (e.g., DateTime).  But I don't want to introduce
+    # any CPAN dependencies here, so just do sometime simple, even if
+    # it's a bit laborious. Use a few private helper functions for
+    # this kind of functionality.
+
+    sub _get_timestamp {
+        my $filename = shift;
+
+        my $ret;
+        if (-x $filename) {
+            my $out = `$filename --version`;
+            $out =~ m/GNU config\.[a-z]+ \((.+)\)/;
+            $ret = $1;
+        }
+
+        return $ret;
+    }
+
+    sub _split_timestamp {
+        my $ts = shift;
+
+        $ts =~ m/(\d\d\d\d)-(\d\d)-(\d\d)/;
+        return $1, $2, $3;
+    }
+
+    # Returns true if timestamp $a > timestamp $b.
+    sub _timestamp_gt {
+        my ($a, $b) = @_;
+
+        my ($year_a, $month_a, $day_a) = _split_timestamp($a);
+        my ($year_b, $month_b, $day_b) = _split_timestamp($b);
+
+        # Don't try to be clever -- just do a simple set of explicit
+        # comparisons.
+        if ($year_a > $year_b) {
+            return 1;
+        } elsif ($year_a < $year_b) {
+            return 0;
+        } else {
+            if ($month_a > $month_b) {
+                return 1;
+            } elsif ($month_a < $month_b) {
+                return 0;
+            } else {
+                if ($day_a > $day_b) {
+                    return 1;
+                } else {
+                    return 0;
+                }
+            }
+        }
+    }
+
+    my ($topdir) = @_;
+
+    # Find the stashed known-good files, and get their version
+    # timestamps.
+    my $cached_dir = "$topdir/config/from-savannah";
+    my @files = qw/config.guess config.sub/;
+    foreach my $file (@files) {
+        if (!exists($cached_config_timestamps{$file})) {
```

Comment:
this seems overly complicated.  why not just hard fail if you can't find the two autogen files, and then you don't need the cache, since you can just get the new and old timestamp and compare?

Reply:
Sure, I can do that.

I actually initially had this subroutine called in each directory -- so caching the values was important.  Now that I'm no longer doing that (i.e., I'm only calling it once at the top level), you're right: this is a little overly complicated.